File: 885/solution.cpp
class Solution {
public:
    vector<vector<int>> spiralMatrixIII(int rows, int cols, int rStart, int cStart) {

        vector<vector<int>> dirs = {{0,1},{1,0},{0,-1},{-1,0}};
        vector<vector<int>> res = {};
        int steps = 1;
        int r = rStart;
        int c = cStart;
        int i = 0;
        while (res.size() < rows * cols) {
            for (int x = 0; x < 2; x++) {
                vector<int> curDir = dirs[i];
                for (int y = 0; y < steps; y++) {
                    if (r >= 0 && r < rows && c >= 0  && c < cols) {
                        res.push_back({r, c});
                    }
                    r = r + curDir[0];
                    c = c + curDir[1];
                }
                i = (i + 1) % 4;
            }
            steps++;
        }
        return res;
    }
};

File: 885/solution.js
/**
 * @param {number} rows
 * @param {number} cols
 * @param {number} rStart
 * @param {number} cStart
 * @return {number[][]}
 */

// took too long

var spiralMatrixIII = function(rows, cols, rStart, cStart) {
  const outputLength = rows * cols;
  const dirs = ['East','South', 'West', 'North'];
  let currentDirection = dirs[0]
  let output = []
  let x = cStart
  let y = rStart
  let spiralCounter = 1
  let directionCounter = 0
  output.push([x, y])
  while (output.length <= outputLength) {
      if (currentDirection == dirs[0]) {
          // East
          x++
          directionCounter++;
          if (spiralCounter == directionCounter) {
              currentDirection = dirs[1]
              directionCounter = 0
          }
          if (0 >= x && x < cols && 0 >= y && y < rows) {
              output.push([x, y])
          }
      }
      if (currentDirection == dirs[1]) {
          y++
          directionCounter++
          if (spiralCounter == directionCounter) {
              currentDirection = dirs[2]
              directionCounter = 0
              spiralCounter++
          }
          if (0 >= x && x < cols && 0 >= y && y < rows) {
              output.push([x, y])
          }
      }
      if (currentDirection == dirs[2]) {
          x--
          directionCounter++
          if (spiralCounter == directionCounter) {
              currentDirection = dirs[3]
              directionCounter = 0
          }
          if (0 >= x && x < cols && 0 >= y && y < rows) {
              output.push([x, y])
          }
      }
      if (currentDirection == dirs[3]) {
          y++
          directionCounter++
          if (spiralCounter == directionCounter) {
              currentDirection = dirs[0]
              directionCounter = 0
              spiralCounter++
          }
          if (0 >= x && x < cols && 0 >= y && y < rows) {
              output.push([x, y])
          }
      }
  }
  return output
};